terraform/context.go

```diff
@@ -892,7 +906,15 @@ func (c *Context) walk(
 	return walker, realErr
 }
 
-func (c *Context) watchStop(walker *ContextGraphWalker, stopCh, doneCh <-chan struct{}) {
+func (c *Context) watchStop(walker *ContextGraphWalker, doneCh <-chan struct{}) {
+	// Get the stop channel. runContext might be nil only during tests.
+	// If this is called during a proper run operation, this will never
+	// be nil.
+	var stopCh <-chan struct{}
+	if ctx := c.runContext; ctx != nil {
```

Comment:
This definitely races with `acquireRun`, which sets the context fields under a mutex. What we can do here is have `walk` get the `stopCh` and pass it in to avoid reading that field without the mutex:

    func (c *Context) watchStop(walker *ContextGraphWalker, stopCh, doneCh <-chan struct{})